### ISSUE TITLE
Standardize Joint Constructors

### ISSUE THREAD

**Issue by mxgrey** [dartsim/dart#280]
It would be nice if the constructors for different Joint types were standardized to make templated factories doable. For example if I wanted a function like 

```c++
template<JointType>
Joint* createJoint(const std::string& name, /* Other parameters like offset transforms and stuff */)
{
  Joint* joint = new JointType(name);
  /* Apply the different parameters */
  return joint;
}
```

This currently cannot be done for all joint types since RevoluteJoint and PrismaticJoint have a different style of constructor than many of the others (taking an axis argument before a name argument). Of course this can be worked around by just using the default constructor in the factory and then changing the name later, but I wonder if we can abstract the concept of an "axis" so that all the different Joint constructors can have compatible signatures.

Essentially, I think Joint types should try to share a template concept as much as possible.

**Comment by mxgrey** [comment]
This issue has been resolved in #369

**Comment by mxgrey** [comment]
It might be worth mentioning the solution that we arrived at, just in case someone looks over this thread later:

We are using a nested ``Properties`` class which gets passed into the constructor of each Joint type. So each unique Joint type must define its own ``Properties`` class. For example:

```
RevoluteJoint::Properties
FreeJoint::Properties
TranslationalJoint::Properties
UniversalJoint::Properties
// etc ...
```

all exist. This allows us to have a template-based Joint factory in the Skeleton class where a single function can handle the construction of any Joint type, *including* any Joint types that might get created by an end-user. The only condition to that is the end-user needs to make sure that their custom Joint type has its own Properties class defined. The end-user should also be sure to define setProperties, getProperties, copy, and clone functions to ensure that their custom Joint type is fully compatible with the DART API. The native Joint types in DART can be used as an example of how to set up a custom Joint type. RevoluteJoint probably offers the most straightforward implementation example.

This approach has also been applied to BodyNode and SoftBodyNode.

The factory function that creates Joints and BodyNodes is Skeleton::createJointAndBodyNodePair
